cpp/ql/src/Likely Bugs/Likely Typos/ExprHasNoEffect.ql

```diff
@@ -84,6 +84,7 @@ where
   not peivc.getEnclosingFunction().isDefaulted() and
   not exists(Macro m | peivc = m.getAnInvocation().getAnExpandedElement()) and
   not peivc.isFromTemplateInstantiation(_) and
```

Comment:
Is this line still contributing positively?

Reply:
Yes - it handles cases where a call in a void context has a pure target in some instantiations and a target with side effects in other instantiations. Ideally we'd handle that with a forall, but I don't think we have any way of matching the corresponding elements in different instantiations.

Reply:
The unusual thing here is that we're excluding both uninstantiated templates _and_ instantiated templates. That means we're excluding all template code. We usually exclude either one or the other, but in this case there has been false positives in both types of code.

We could try something elaborate like counting the number of potential alerts in the template and its instantiations and reporting an alert only if those counts match, but I don't think this query is important enough for us to spend time on that.

Reply:
OK, sounds like this is a reasonable solution then.